homeassistant/components/recorder/__init__.py

```diff
@@ -191,38 +199,25 @@ def run(self):
                 self.queue.task_done()
                 return
 
-            elif event.event_type == EVENT_TIME_CHANGED:
+            if event.event_type == EVENT_TIME_CHANGED:
                 self.queue.task_done()
                 continue
 
             session = Session()
             dbevent = Events.from_event(event)
             session.add(dbevent)
-
-            for _ in range(0, RETRIES):
-                try:
-                    session.commit()
-                    break
-                except sqlalchemy.exc.OperationalError as e:
-                    log_error(e, retry_wait=QUERY_RETRY_WAIT,
-                              rollback=True)
+            self._commit(session)
 
             if event.event_type != EVENT_STATE_CHANGED:
                 self.queue.task_done()
                 continue
 
+            # Can possibly be combined into single session with the above
             session = Session()
             dbstate = States.from_event(event)
-
-            for _ in range(0, RETRIES):
-                try:
-                    dbstate.event_id = dbevent.event_id
-                    session.add(dbstate)
-                    session.commit()
-                    break
-                except sqlalchemy.exc.OperationalError as e:
-                    log_error(e, retry_wait=QUERY_RETRY_WAIT,
-                              rollback=True)
+            dbstate.event_id = dbevent.event_id
+            session.add(dbstate)
```

Comment:
Thanks @rhooper, taking into account that you have to redo work, your original method makes sense to me now! :smile:

Original code had 4 commits, with 1 retrying & redoing work. Will cover all 4 using chosen option:

- (1) Original `for ... RETRIES` clauses
- (2) `retry_work(the_work_callback)`

Your & @balloob's final call.